hail/src/main/scala/is/hail/HailLSM.scala

```diff
@@ -0,0 +1,106 @@
+package is.hail
+
+import is.hail.annotations._
+import is.hail.utils._
+import is.hail.expr.types.physical.PType
+import is.hail.io.TypedCodecSpec
+import java.io.{ DataInput, DataOutput, EOFException, File, IOException, InputStream, OutputStream }
+import com.indeed.lsmtree.core._
+import com.indeed.util.serialization._
+import com.indeed.util.serialization.array._
+import java.util.Comparator
+import java.io.DataInputStream
+import java.io.DataOutputStream
+
+class DataOutputIsOutputStream(
+  var out: DataOutput
+) extends OutputStream {
+  override def close() { }
```

Comment:
This definitely feels wrong.  Where is the DataOutput that's not a stream coming from?  Maybe the consumer of `DataOutputIsOutputStream` should just be expecting a DataOutput, and not a stream?

Reply:
Indeed's LSM uses an in-memory cache of the serialized data. Their (de)serialization interface takes `DataOutput`  not `OutputStream`. They want to serialize a value to a byte array. ByteArrayOutputStream is not a DataOutput, so instead they use [ByteArrayDataOutput](https://guava.dev/releases/13.0/api/docs/com/google/common/io/ByteArrayDataOutput.html), which does not implement `OutputStream`.

Hail's (de)serialization interface takes `OutputStream`. This is effectively a fallback bridge between Indeed and Hail's (de)serialization interfaces. AFAIK, it is only used for this in-memory cache. The disk `DataOutput` objects are instances of `DataOutputStream`, so I cast them to `OutputStream` below.